src/gf.c

```diff
@@ -189,88 +141,84 @@ void jl_mk_builtin_func(jl_datatype_t *dt, const char *name, jl_fptr_args_t fptr
         jl_set_const(jl_core_module, sname, f);
         dt = (jl_datatype_t*)jl_typeof(f);
     }
-    jl_method_instance_t *li = jl_new_method_instance_uninit();
-    li->invoke = jl_fptr_args;
-    li->specptr.fptr1 = fptr;
-    li->specTypes = (jl_value_t*)jl_anytuple_type;
-    li->min_world = 1;
-    li->max_world = ~(size_t)0;
 
-    JL_GC_PUSH1(&li);
     jl_method_t *m = jl_new_method_uninit(jl_core_module);
-    li->def.method = m;
-    jl_gc_wb(li, m);
     m->name = sname;
     m->module = jl_core_module;
     m->isva = 1;
     m->nargs = 2;
     m->sig = (jl_value_t*)jl_anytuple_type;
     m->slot_syms = jl_an_empty_string;
 
+    JL_GC_PUSH1(&m);
+    jl_method_instance_t *mi = jl_get_specialized(m, (jl_value_t*)jl_anytuple_type, jl_emptysvec);
+    m->unspecialized = mi;
+    jl_gc_wb(m, mi);
+
+    jl_code_instance_t *codeinst = jl_set_method_inferred(mi, (jl_value_t*)jl_any_type, jl_nothing, jl_nothing,
+        0, 1, ~(size_t)0);
+    codeinst->specptr.fptr1 = fptr;
+    codeinst->invoke = jl_fptr_args;
+
     jl_methtable_t *mt = dt->name->mt;
     jl_typemap_insert(&mt->cache, (jl_value_t*)mt, jl_anytuple_type,
-        NULL, jl_emptysvec, (jl_value_t*)li, 0, &lambda_cache, 1, ~(size_t)0, NULL);
+        NULL, jl_emptysvec, (jl_value_t*)mi, 0, &lambda_cache, 1, ~(size_t)0, NULL);
     JL_GC_POP();
 }
 
-// run type inference on lambda "li" for given argument types.
-// returns the inferred source, and may cache the result in li
-// if successful, also updates the li argument to describe the validity of this src
+// run type inference on lambda "mi" for given argument types.
+// returns the inferred source, and may cache the result in mi
+// if successful, also updates the mi argument to describe the validity of this src
 // if inference doesn't occur (or can't finish), returns NULL instead
-jl_code_info_t *jl_type_infer(jl_method_instance_t **pli JL_ROOTS_TEMPORARILY, size_t world, int force)
+jl_code_info_t *jl_type_infer(jl_method_instance_t *mi, size_t world, int force)
 {
     JL_TIMING(INFERENCE);
     if (jl_typeinf_func == NULL)
         return NULL;
+    if (jl_is_method(mi->def.method) && mi->def.method->unspecialized == mi)
```

Comment:
If nospecialize is set on all arguments, will we still get a MethodInstance here to infer? It is useful to at least resolve things like loops and literals.

Reply:
That's a nice feature that this work will allow us to add. Since we don't have it now, I'm not concerned about it right now.